canopen/source/co_csdo.c

```diff
@@ -450,11 +495,392 @@ void COCSdoTransferFinalize(CO_CSDO *csdo) {
+                                uint32_t timeout){
+    CO_IF_FRM frm;
+    uint32_t ticks;
+    
+     if ((csdo == 0) || (callback == 0) ||
+        (buf  == 0) || (size     == 0)) {
+        /*
+         * Invalid argument
+         *
+         * TODO: Maybe we can tolerate NULL callback,
+         * but i do not see any reason why request
+         * without callback should be triggered.
+         */
+        return CO_ERR_BAD_ARG;
+    }
+    if (csdo->State == CO_CSDO_STATE_INVALID) {
+        /* Requested SDO client is disabled */
+        return CO_ERR_SDO_OFF;
+    }
+    if (csdo->State == CO_CSDO_STATE_BUSY) {
+        /* Requested SDO client is busy */
+        return CO_ERR_SDO_BUSY;
+    }
+
+    /*
+     * Set client as busy to prevent its usage
+     * until requested transfer is complete
+     */
+    csdo->State = CO_CSDO_STATE_BUSY;
+
+     /* Update transfer info */
+    csdo->Tfer.Type  = CO_CSDO_TRANSFER_UPLOAD_SEGMENT;
+    csdo->Tfer.Abort = 0;
+    csdo->Tfer.Idx   = CO_GET_IDX(key);
+    csdo->Tfer.Sub   = CO_GET_SUB(key);
+    csdo->Tfer.Buf   = buf;
+    csdo->Tfer.Size  = size;
+    csdo->Tfer.Tmt   = timeout;
+    csdo->Tfer.Call  = callback;
+    csdo->Tfer.Buf_Idx = 0;
+
+    /* Transmit transfer initiation directly */
+    CO_SET_ID  (&frm, csdo->TxId       );
+    CO_SET_DLC (&frm, 8                );
+    CO_SET_BYTE(&frm, 0x40          , 0);
+    CO_SET_WORD(&frm, csdo->Tfer.Idx, 1);
+    CO_SET_BYTE(&frm, csdo->Tfer.Sub, 3);
+    CO_SET_LONG(&frm, 0,              4);
+
+    ticks = COTmrGetTicks(&(csdo->Node->Tmr), timeout, CO_TMR_UNIT_1MS);
+    csdo->Tfer.Tmr = COTmrCreate(&(csdo->Node->Tmr), ticks, 0, &COCSdoTimeout, csdo);
+
+    (void)COIfCanSend(&csdo->Node->If, &frm);
+
+    return CO_ERR_NONE;
+
+}
+
+
+
+CO_ERR COCSdoInitDownloadSegmented(CO_CSDO *csdo)
+{
+    CO_ERR result = CO_ERR_SDO_SILENT;
+    uint32_t ticks;
+    uint16_t Idx;
+    uint8_t Sub;
+    uint8_t n, width;
+    uint8_t c_bit = 0;
+    uint8_t cmd;
+    CO_IF_FRM frm;
+
+    Idx = CO_GET_WORD(csdo->Frm, 1);
+    Sub = CO_GET_BYTE(csdo->Frm, 3);
+
+    if (Idx == csdo->Tfer.Idx &&
+        Sub == csdo->Tfer.Sub)
+    {
+        CO_SET_ID  (&frm, csdo->TxId       );
+        CO_SET_DLC (&frm, 8                );
+
+        /* clean frm data */
+        CO_SET_LONG(&frm, 0, 0);
+        CO_SET_LONG(&frm, 0, 4);
+
+        width = csdo->Tfer.Size - csdo->Tfer.Buf_Idx;
+
+        if (width > 7)
+        {
+            width = 7;
+        }
+        else
+        {
+            c_bit = 1;
+        }
+        
+        for (n = 1; n <= width; n++)
+        {
+            CO_SET_BYTE(&frm, csdo->Tfer.Buf[csdo->Tfer.Buf_Idx], n);
+            csdo->Tfer.Buf_Idx++;
+        }
+
+        cmd = (uint8_t)(csdo->Tfer.TBit << 4) |
+              (uint8_t)(((7 - width) << 1)) | 
+              (uint8_t)(c_bit);
+
+        CO_SET_BYTE(&frm, cmd, 0);
+
+         
+        csdo->Tfer.TBit = 0;
```

Comment:
We should move line #729 above line #722.
Otherwise, we use a random stack value in the calculation of `cmd`.